dice-where-downloader-lib/src/main/java/technology/dice/dicewhere/downloader/stream/StreamWithMD5Decorator.java

```diff
@@ -1,41 +1,82 @@
 package technology.dice.dicewhere.downloader.stream;
 
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
 import java.io.IOException;
 import java.io.InputStream;
 import java.security.DigestInputStream;
 import java.security.MessageDigest;
 import java.security.NoSuchAlgorithmException;
+import java.util.Optional;
 import javax.xml.bind.annotation.adapters.HexBinaryAdapter;
 import technology.dice.dicewhere.downloader.md5.MD5Checksum;
 
 public class StreamWithMD5Decorator extends InputStream {
 
+  private static final int BUFFER_SIZE = 8192;
+  private static final HexBinaryAdapter HEX_BINARY_ADAPTER = new HexBinaryAdapter();
+
   private final MessageDigest md5;
-  DigestInputStream inputStream;
+  private final DigestInputStream originalStream;
+
+  private Optional<InputStream> duplicatedStream = Optional.empty();
+  private Optional<MD5Checksum> md5Checksum = Optional.empty();
 
-  private StreamWithMD5Decorator(DigestInputStream inputStream, MessageDigest md5) {
-    this.inputStream = inputStream;
+  private StreamWithMD5Decorator(DigestInputStream originalStream, MessageDigest md5)
```

Comment:
This class is lying and likely to cause issues. One of the major benefits of `InputStream`s is that they do not require the whole contents of the file (or other source) to be read into memory at any time. But this class subverts this as the passed stream is consumed and written to a temporary array - this is not what clients of this class would expect.

Reply:
Yes, we're indeed copying an InputStream for origin - 
But if the stream is not read, I cannot generate an hash from it  ( gives me `d41d8cd98f00b204e9800998ecf8427e` checksum for an empty input )
and if the stream is read it can't be re-used. 
So this implementation tries to perform the md5 whilst letting you a InputStream to be used elsewhere